apps/storeutl.c

```diff
@@ -0,0 +1,185 @@
+        BIO_printf(bio_err, "%s: Unknown extra parameters after URI\n", prog);
+        goto opthelp;
+    }
+
+    if (!app_passwd(passinarg, NULL, &passin, NULL)) {
+        BIO_printf(bio_err, "Error getting passwords\n");
+        goto end;
+    }
+    pw_cb_data.password = passin;
+    pw_cb_data.prompt_info = argv[0];
+
+    out = bio_open_default(outfile, 'w', FORMAT_TEXT);
+    if (out == NULL)
+        goto end;
+
+    if ((store_ctx = STORE_open(argv[0], get_ui_method(), &pw_cb_data, NULL,
+                                   NULL)) == NULL) {
+        BIO_printf(bio_err, "Couldn't open file or uri %s\n", argv[0]);
+        ERR_print_errors(bio_err);
+        goto end;
+    }
+
+    /* From here on, we count errors, and we'll return the count at the end */
+    ret = 0;
+
+    while (!STORE_eof(store_ctx)) {
+        STORE_INFO *info = STORE_load(store_ctx);
+
+        if (info == NULL) {
+            ERR_print_errors(bio_err);
+            ret++;
+            break;
+        } else if (STORE_INFO_get_type(info) == STORE_INFO_UNSPECIFIED) {
+            goto cont;
+        } else if (STORE_INFO_get_type(info) == STORE_INFO_NAME) {
+            BIO_printf(bio_out, "%d: %s: %s\n", items,
+                       STORE_INFO_type_string(STORE_INFO_get_type(info)),
+                       STORE_INFO_get0_NAME(info));
+            if (STORE_INFO_get0_NAME_description(info))
+                BIO_printf(bio_out, "%s\n",
+                           STORE_INFO_get0_NAME_description(info));
+        } else {
+            BIO_printf(bio_out, "%d: %s\n", items,
+                       STORE_INFO_type_string(STORE_INFO_get_type(info)));
+        }
+
+        if (info != NULL) {
+            /*
+             * Unfortunately, PEM_X509_INFO_write_bio() is sorely lacking in
+             * functionality, so we must figure out how exactly to write things
+             * ourselves...
+             */
+            switch (STORE_INFO_get_type(info)) {
+            case STORE_INFO_NAME:
+                break;
+            case STORE_INFO_PARAMS:
+                if (text)
+                    EVP_PKEY_print_params(out, STORE_INFO_get0_PARAMS(info),
+                                          0, NULL);
+                if (!noout)
+                    PEM_write_bio_Parameters(out, STORE_INFO_get0_PARAMS(info));
+                break;
+            case STORE_INFO_PKEY:
+                if (text)
+                    EVP_PKEY_print_private(out, STORE_INFO_get0_PKEY(info),
+                                           0, NULL);
+                if (!noout)
+                    PEM_write_bio_PrivateKey(out, STORE_INFO_get0_PKEY(info),
+                                             NULL, NULL, 0, NULL, NULL);
+                break;
+            case STORE_INFO_CERT:
+                if (text)
+                    X509_print(out, STORE_INFO_get0_CERT(info));
+                if (!noout)
+                    PEM_write_bio_X509(out, (X509 *)STORE_INFO_get0_CERT(info));
```

Comment:
Because `PEM_write_bio_X509` takes a non-const and `STORE_INFO_get0_CER` returns a const.  It's a bit of a cheat, but we know for a fact that the X509 can't be in const space, so...